### PR TITLE
Integrate Data Balance insights into the RAI Dashboard

### PR DESCRIPTION
**Pull request by ms-kashyap**
## Description

Our team at Microsoft believes that having a well balanced data representation is critical when developing models in a responsible way, specially in terms of fairness. Finding data balance insights in the data exploration stage reduces risks for unbalanced models that are pushed to production, such as [this one](https://www.reuters.com/article/us-amazon-com-jobs-automation-insight-idUSKCN1MK08G).

We added the ability to compute data balance insights on both [Spark](https://microsoft.github.io/SynapseML/docs/features/responsible_ai/Data%20Balance%20Analysis/) and [Python](https://pypi.org/project/raimitigations/).

However, these insights are hard to understand when printed out in a Python notebook. Therefore, we would like to incorporate these data balance insights as part of the Responsible AI Dashboard, specifically the Data Explorer part.

These visualizations do not include mitigation steps yet, but this is what we will work on next. To learn more about these measures, [please click here](https://microsoft.github.io/SynapseML/docs/features/responsible_ai/Data%20Balance%20Analysis/#measure-explanations).

---

Please note, the following things are still to be done, but I wanted to get this Pull Request out for review while I work on them:
- Localization
- Tests
- Documentation
- Extra notebook showing how the feature works with Spark data
- Finalization of the Python `DataBalanceManager`

## Areas changed

npm packages changed:

- [ ] responsibleai/causality
- [x] responsibleai/core-ui
- [ ] responsibleai/counterfactuals
- [x] responsibleai/dataset-explorer
- [ ] responsibleai/fairness
- [ ] responsibleai/interpret
- [ ] responsibleai/localization
- [ ] responsibleai/mlchartlib
- [x] responsibleai/model-assessment

Python packages changed:

- [ ] raiwidgets
- [x] responsibleai
- [ ] erroranalysis
- [ ] rai_core_flask

## Tests

(As mentioned above, I will work on tests while this PR is out for review)

- [ ] No new tests required.
- [ ] New tests for the added feature are part of this PR.
- [ ] I validated the changes manually.

## Screenshots (if appropriate):

Overall view:
Overall-View

Feature Balance Measures (gif):
Feature-Balance-Measures

Distribution Balance Measures (gif):
Distribution-Balance-Measures

Aggregate Balance Measures (gif):
Aggregate-Balance-Measures

## Documentation:

- [X] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: apps/dashboard/src/model-assessment/__mock_data__/adultCensus.ts [hunk 1]
```diff
@@ -25,6 +25,1475 @@ export const adultCensusWithFairnessDataset: IDataset = {
     "native-country"
   ],
   class_names: ["<=50K", ">50K"],
+  data_balance_measures: {
```
Comment (imatiach-msft) [inline apps/dashboard/src/model-assessment/__mock_data__/adultCensus.ts:28]: hmm I see errors like "Identifier 'data_balance_measures' is not in camel case"
I wonder if you need to make these camel case then, like dataBalanceMeasures.  It seems like some variables like class_names are not camel case though.  I'm not sure why it doesn't complain about them.  Perhaps it's whitelisted somewhere.

Comment (imatiach-msft) [inline apps/dashboard/src/model-assessment/__mock_data__/adultCensus.ts:28]: hmm I see errors like "Identifier 'data_balance_measures' is not in camel case"
I wonder if you need to make these camel case then, like dataBalanceMeasures.  It seems like some variables like class_names are not camel case though.  I'm not sure why it doesn't complain about them.  Perhaps it's whitelisted somewhere.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -48,6 +48,30 @@ All pull requests need to abide by the following criteria to be accepted:
 - support for localization in code (translations need not be provided)
 - tests for added / changed functionality
 
+## Working with Git
```
Comment (imatiach-msft) [inline CONTRIBUTING.md:51]: I feel like this whole contributing.md change could just be its own PR, this PR is already really large as it is

Comment (imatiach-msft) [inline CONTRIBUTING.md:51]: I feel like this whole contributing.md change could just be its own PR, this PR is already really large as it is

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -48,6 +48,30 @@ All pull requests need to abide by the following criteria to be accepted:
...
+
+It is a good idea to fork the `responsible-ai-toolbox` repo and work inside your own fork so that the main repo does not have to deal with so many branches and commits.
+
+### First Time Setup
+
+- Click on the `Fork` button in the top right corner of the [`responsible-ai-toolbox` repo.](https://github.com/microsoft/responsible-ai-toolbox) to create a fork.
+- Clone your forked repo to your local machine. Git will refer to your forked repo as `origin`.
+- Run `git remote add upstream https://github.com/microsoft/responsible-ai-toolbox.git` to register the upstream repo as `upstream`.
```
Comment (imatiach-msft) [inline CONTRIBUTING.md:59]: you will probably want to use the ssh url instead of "https://github.com", so you won't need to type password in on syncs

Comment (imatiach-msft) [inline CONTRIBUTING.md:59]: you will probably want to use the ssh url instead of "https://github.com", so you won't need to type password in on syncs

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -48,6 +48,30 @@ All pull requests need to abide by the following criteria to be accepted:
...
+
+To pull the latest changes from the `main` branch in the upstream repo to your branch:
+
+### Merging latest changes into your branch
+
+Follow this part only when your branch has become stale and would like to pull the latest changes into your branch.
```
Comment (imatiach-msft) [inline CONTRIBUTING.md:65]: nit:
and would -> and you would

I think this is more correct?

Comment (imatiach-msft) [inline CONTRIBUTING.md:65]: nit:
and would -> and you would

I think this is more correct?

File: libs/core-ui/src/lib/Interfaces/DataBalanceInterfaces.ts [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+// The key is the formatted name instead of var name because its keys are used to populate the Measure Dropdown
+export const ApprovedFeatureBalanceMeasures = new Map<
+  string,
+  FeatureBalanceMeasureInfo
+>([
+  [
+    "Demographic Parity",
+    {
+      description:
+        "<b>Demographic Parity</b> measures how much one class receives the positive outcome compared to the other class.<br>As close to 0 means both classes receive the positive outcome equally. A <b>positive value</b> means the first class<br>sees the positive label more than the second class. A <b>negative value</b> means the opposite.",
```
Comment (imatiach-msft) [inline libs/core-ui/src/lib/Interfaces/DataBalanceInterfaces.ts:77]: As close to 0 means -> A value close to zero

File: libs/core-ui/src/lib/Interfaces/DataBalanceInterfaces.ts [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+// The key is the formatted name instead of var name because its keys are used to populate the Measure Dropdown
+export const ApprovedFeatureBalanceMeasures = new Map<
+  string,
+  FeatureBalanceMeasureInfo
+>([
+  [
+    "Demographic Parity",
+    {
+      description:
```
Comment (imatiach-msft) [inline libs/core-ui/src/lib/Interfaces/DataBalanceInterfaces.ts:76]: all of these descriptions need to go into the localization files

Comment (imatiach-msft) [reply libs/core-ui/src/lib/Interfaces/DataBalanceInterfaces.ts:76]: since you do this in line highlighting:
<b>Demographic Parity</b>
I wonder if you should break apart the terms in the localization.  Not sure how this will work with different languages.  Perhaps it might be better to remove the highlights in text.

File: libs/core-ui/src/lib/Interfaces/DataBalanceInterfaces.ts [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+        "<b>Demographic Parity</b> measures how much one class receives the positive outcome compared to the other class.<br>As close to 0 means both classes receive the positive outcome equally. A <b>positive value</b> means the first class<br>sees the positive label more than the second class. A <b>negative value</b> means the opposite.",
+      range: [-1, 1],
+      varName: "dp"
+    }
+  ],
+  [
+    "Pointwise Mutual Information (PMI)",
+    {
+      // range: [-inf, +inf] => not specified so chart uses min/max values
+      description:
```
Comment (imatiach-msft) [inline libs/core-ui/src/lib/Interfaces/DataBalanceInterfaces.ts:86]: similar comment for all of these other descriptions, they need to be localized

File: libs/core-ui/src/lib/util/getDataBalanceChartOptions.ts [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+      if (featureValue && colIndex !== rowIndex) {
+        // Add the feature value for comparing class A (col index) to class B (row index)
+        // The feature value may not be the same as the feature value when comparing class B to class A (swapped row and col)
+        data.push([colIndex, rowIndex, featureValue]);
+      }
+    });
+  });
+
+  const titleOptions = {
+    text: `Comparing ${selectedMeasure} on all classes of ${selectedFeature}`
```
Comment (imatiach-msft) [inline libs/core-ui/src/lib/util/getDataBalanceChartOptions.ts:57]: these need to be localized as mentioned above

File: libs/core-ui/src/lib/util/getDataBalanceChartOptions.ts [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+        data,
+        dataLabels: {
+          enabled: true
+        },
+        name: selectedFeature,
+        type: "heatmap"
+      }
+    ],
+    subtitle: {
+      text: `<a target="_blank" href="https://microsoft.github.io/SynapseML/docs/features/responsible_ai/Data%20Balance%20Analysis/#feature-balance-measures">Reference Link for Feature Balance Measures</a>`,
```
Comment (imatiach-msft) [inline libs/core-ui/src/lib/util/getDataBalanceChartOptions.ts:108]: localize

File: libs/core-ui/src/lib/util/getDataBalanceChartOptions.ts [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+  } as TitleOptions;
+  if (datasetName) {
+    titleOptions.text += ` in ${datasetName}`;
+  }
+
+  const colorAxisOptions = {} as ColorAxisOptions;
+  if (measure.range) {
+    [colorAxisOptions.min, colorAxisOptions.max] = measure.range;
+  } else {
+    const featureValues = data.map((row) => row[2]);
```
Comment (imatiach-msft) [inline libs/core-ui/src/lib/util/getDataBalanceChartOptions.ts:67]: I wonder if a reduce on data which calculates the min/max in a tuple might be slightly cleaner and more performant since it would loop just once instead of three times over the data.  But maybe this is written in a clearer way anyway.  Not sure.

File: libs/core-ui/src/lib/util/getDataBalanceChartOptions.ts [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+  });
+
+  const titleOptions = {
+    text: `Comparing ${features.join(", ")} with Uniform Distribution`
+  } as TitleOptions;
+  if (datasetName) {
+    titleOptions.text += ` in ${datasetName}`;
+  }
+
+  // TODO: Show the measure's description in a tooltip. Things tried but failed to work:
```
Comment (imatiach-msft) [inline libs/core-ui/src/lib/util/getDataBalanceChartOptions.ts:196]: could you explain this a bit more?  Did you use the tooltip from fabricui?

File: libs/core-ui/src/lib/util/getDataBalanceChartOptions.ts [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+  // - Overriding the default tooltip with a custom one (worked but could not render the colored circles that default tooltip has)
+  // - Adding a custom hover-over as a load() event (could not figure out which measure was being hovered over)
+  return {
+    chart: {
+      numberFormatter: (value: number) => value.toFixed(3),
+      type: "column"
+    },
+    series: data,
+    subtitle: {
+      text: `<a target="_blank" href="https://microsoft.github.io/SynapseML/docs/features/responsible_ai/Data%20Balance%20Analysis/#distribution-balance-measures">Reference Link for Distribution Balance Measures</a>`,
```
Comment (imatiach-msft) [inline libs/core-ui/src/lib/util/getDataBalanceChartOptions.ts:207]: It's weird for us to link to the SynapseML docs from this repository.  I wonder if this page can be replicated in the RAI toolbox documentation website.

File: libs/core-ui/src/lib/util/getDataBalanceChartOptions.ts [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+      useHTML: true
+    },
+    title: titleOptions,
+    xAxis: {
+      categories: measureOptions,
+      title: { text: "Distribution Measure" }
```
Comment (imatiach-msft) [inline libs/core-ui/src/lib/util/getDataBalanceChartOptions.ts:213]: localization

File: libs/core-ui/src/lib/util/getDataBalanceChartOptions.ts [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+    plotOptions: {
+      bar: {
+        dataLabels: {
+          enabled: true
+        }
+      }
+    },
+    series: data,
+    subtitle: {
+      text: `<a target="_blank" href="https://microsoft.github.io/SynapseML/docs/features/responsible_ai/Data%20Balance%20Analysis/#aggregate-balance-measures">Reference Link for Aggregate Balance Measures</a>`,
```
Comment (imatiach-msft) [inline libs/core-ui/src/lib/util/getDataBalanceChartOptions.ts:264]: same comment as above, perhaps we can have a different link to our docs here

File: libs/dataset-explorer/src/lib/DataBalanceTab.tsx [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+          {
+            "This tab requires the dataset to contain already computed data balance measures." // TODO: Replace with localization
+          }
+        </MissingParametersPlaceholder>
+      );
+    }
+
+    // If some types of measures are available but others are not (i.e. feature balance measures are not available but
+    // distribution balance measures are), we still display the measures that are available.
+    const dataBalanceMeasures = this.context.dataset.data_balance_measures,
```
Comment (imatiach-msft) [inline libs/dataset-explorer/src/lib/DataBalanceTab.tsx:44]: This seems like a weird way to define these constants, can each be put on a separate line instead of all on one line?  Is this type of style recommended?  I haven't seen it before in this repository.

File: libs/dataset-explorer/src/lib/DataBalanceTab.tsx [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+      featureBalanceMeasures = dataBalanceMeasures.featureBalanceMeasures,
+      distributionBalanceMeasures =
+        dataBalanceMeasures.distributionBalanceMeasures,
+      aggregateBalanceMeasures = dataBalanceMeasures.aggregateBalanceMeasures;
+
+    return (
+      <div className={styles.page}>
```
Comment (imatiach-msft) [inline libs/dataset-explorer/src/lib/DataBalanceTab.tsx:51]: we try to use the Stack from office fabricui instead of div in this repo when possible, would it be possible to replace these with Stack?

Comment (imatiach-msft) [reply libs/dataset-explorer/src/lib/DataBalanceTab.tsx:51]: hmm, actually, there is only a single div here.  Maybe that's ok.  I wonder if using Stack and specifying tokens might help remove the br though

File: libs/dataset-explorer/src/lib/DefaultTab.tsx [hunk 17]
```diff
@@ -0,0 +1,389 @@
...
+
+    const selectedCohortIndex =
+      this.state.selectedCohortIndex >= this.context.errorCohorts.length
+        ? 0
+        : this.state.selectedCohortIndex;
+
+    const plotlyProps = generatePlotlyProps(
+      this.context.jointDataset,
+      this.state.chartProps,
+      this.context.errorCohorts.map((errorCohort) => errorCohort.cohort)[
```
Comment (imatiach-msft) [inline libs/dataset-explorer/src/lib/DefaultTab.tsx:106]: I think you can remove the map altoghether like this:
```
this.context.errorCohorts[selectedCohortIndex].cohort
```

File: libs/dataset-explorer/src/lib/DefaultTab.tsx [hunk 18]
```diff
@@ -0,0 +1,389 @@
...
+    const yAxisCategories = [
+      ColumnCategories.Index,
+      ColumnCategories.Dataset,
+      ColumnCategories.Outcome
+    ];
+    if (this.state.chartProps.chartType !== ChartTypes.Scatter) {
+      yAxisCategories.push(ColumnCategories.None);
+    }
+    return (
+      <div className={classNames.page}>
```
Comment (imatiach-msft) [inline libs/dataset-explorer/src/lib/DefaultTab.tsx:130]: I see a lot of divs here, I wonder if we can replace these with Stack?

File: libs/dataset-explorer/src/lib/DefaultTab.tsx [hunk 19]
```diff
@@ -0,0 +1,389 @@
...
+  };
+
+  private readonly setColorClose = (): void => {
+    this.setState({ colorDialogOpen: false });
+  };
+
+  private readonly setYOpen = (val: boolean): void => {
+    if (val && this.state.yDialogOpen === false) {
+      this.setState({ yDialogOpen: true });
+      return;
```
Comment (imatiach-msft) [inline libs/dataset-explorer/src/lib/DefaultTab.tsx:324]: Maybe just use if/else instead of doing an early return?  I think it's a bit easier to follow the code that way.
Even better, you can assign a variable like this:
```
const yDialogOpen = val && this.state.yDialogOpen === false
this.setState({ yDialogOpen: yDialogOpen });
```

That reduces 5 lines to 2 lines, and I think it is easier to follow.

File: notebooks/responsibleaidashboard/responsibleaidashboard-census-classification-model-debugging.ipynb [hunk 20]
```diff
@@ -1,523 +1,523 @@
...
-   "mimetype": "text/x-python",
-   "name": "python",
-   "nbconvert_exporter": "python",
-   "pygments_lexer": "ipython3",
-   "version": "3.7.11"
-  }
- },
- "nbformat": 4,
- "nbformat_minor": 5
+  "cells": [
```
Comment (imatiach-msft) [inline notebooks/responsibleaidashboard/responsibleaidashboard-census-classification-model-debugging.ipynb:2]: you will need to add a test for this notebook

File: responsibleai/responsibleai/_interfaces.py [hunk 21]
```diff
@@ -6,8 +6,14 @@
 
 
 class TaskType(str, Enum):
-    CLASSIFICATION = 'classification'
-    REGRESSION = 'regression'
+    CLASSIFICATION = "classification"
```
Comment (imatiach-msft) [inline responsibleai/responsibleai/_interfaces.py:9]: I think the single quotes are preferred over double quotes where possible

File: responsibleai/responsibleai/_internal/constants.py [hunk 22]
```diff
@@ -7,90 +7,95 @@
 class ManagerNames(object):
     """Provide the manager names."""
 
-    CAUSAL = 'causal'
-    COUNTERFACTUAL = 'counterfactual'
-    ERROR_ANALYSIS = 'error_analysis'
-    EXPLAINER = 'explainer'
+    CAUSAL = "causal"
```
Comment (imatiach-msft) [inline responsibleai/responsibleai/_internal/constants.py:10]: strange that the single quotes are converted to double here, they should stay as single quotes

File: responsibleai/responsibleai/managers/data_balance_manager.py [hunk 23]
```diff
@@ -0,0 +1,433 @@
...
+                df=self._df,
+                cols_of_interest=self._cols_of_interest,
+                target_column=self._target_column,
+                pos_label=self._pos_label,
+                backend=self._backend,
+            )
+        )
+
+    def get(self):
+        # TODO docs and implementation
```
Comment (imatiach-msft) [inline responsibleai/responsibleai/managers/data_balance_manager.py:69]: these should be implemented (get and list)

File: responsibleai/responsibleai/managers/data_balance_manager.py [hunk 24]
```diff
@@ -0,0 +1,433 @@
...
+                pos_label=pos_label,
+            )
+        except Exception as e:
+            warnings.warn(
+                f"Failed to compute data balance with pandas due to {e!r}.",
+            )
+
+    @staticmethod
+    def _compute_measures_spark(
+        df: Union[pd.DataFrame, Any],
```
Comment (imatiach-msft) [inline responsibleai/responsibleai/managers/data_balance_manager.py:236]: the any type is strange here, did you mean a spark dataframe?

File: responsibleai/responsibleai/managers/data_balance_manager.py [hunk 25]
```diff
@@ -0,0 +1,433 @@
...
+
+        # Populated in compute()
+        self._data_balance_measures = None
+
+    def add(
+        self,
+        cols_of_interest: List[str],
+        pos_label: Optional[str] = None,
+        custom_data: Optional[Any] = None,
+        backend: Literal["pandas", "spark"] = "pandas",
```
Comment (imatiach-msft) [inline responsibleai/responsibleai/managers/data_balance_manager.py:47]: does spark backend actually work here? If it does, that's very cool!  But then we need to figure out how to test it in this repo.

File: responsibleai/responsibleai/managers/data_balance_manager.py [hunk 26]
```diff
@@ -0,0 +1,433 @@
+# Copyright (c) Microsoft Corporation
```
Comment (imatiach-msft) [inline responsibleai/responsibleai/managers/data_balance_manager.py:1]: I don't see any tests for the python and javascript code.  I think you will need to add a lot of tests for all of the new code.
I would recommend breaking this PR into python and typescript, and maybe get the python in first with all of the necessary tests implemented.

File: responsibleai/responsibleai/managers/data_balance_manager.py [hunk 27]
```diff
@@ -0,0 +1,433 @@
...
+    def name(self):
+        return ManagerNames.DATA_BALANCE
+
+    def _save(self, path):
+        """Abstract method to save the manager.
+
+        :param path: The directory path to save the manager to.
+        :type path: str
+        """
+        # TODO docs and implementation
```
Comment (imatiach-msft) [inline responsibleai/responsibleai/managers/data_balance_manager.py:91]: it's very important to implement the serialization logic, don't forget this part!  Otherwise it won't work with DPv2.

File: responsibleai/responsibleai/managers/data_balance_manager.py [hunk 28]
```diff
@@ -0,0 +1,433 @@
...
+                    " provided to compute data balance measures."
+                )
+            )
+            return False
+
+        if self._backend not in SUPPORTED_BACKENDS:
+            warnings.warn(
+                (
+                    "`backend` must be one of the supported backends:"
+                    f" {SUPPORTED_BACKENDS} to compute data balance measures."
```
Comment (imatiach-msft) [inline responsibleai/responsibleai/managers/data_balance_manager.py:144]: I'm not sure if this is printed in the desired format?
```{SUPPORTED_BACKENDS}```
I think you actually need to do a join on this, otherwise it looks a bit weird to the user?
```
>>> list = ["a", "b", "c"]
>>> print(list)
['a', 'b', 'c']
>>> print(f" {list}")
 ['a', 'b', 'c']
>>> ", ".join(list)
'a, b, c'
```

File: responsibleai/responsibleai/managers/data_balance_manager.py [hunk 29]
```diff
@@ -0,0 +1,433 @@
...
+            )
+
+    @staticmethod
+    def _compute_measures_spark(
+        df: Union[pd.DataFrame, Any],
+        cols_of_interest: List[str],
+        target_column: str,
+        pos_label: str,
+    ) -> Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame]:
+        from synapse.ml import exploratory as SparkDataBalance
```
Comment (imatiach-msft) [inline responsibleai/responsibleai/managers/data_balance_manager.py:241]: I think the imports should still go at the top of the file.  You can just add try/catch around them.

File: responsibleai/responsibleai/rai_insights/rai_insights.py [hunk 30]
```diff
@@ -18,29 +18,30 @@
...
 from responsibleai.utils import _is_classifier
 
-_DATA = 'data'
-_PREDICTIONS = 'predictions'
-_DTYPES = 'dtypes'
-_TRAIN = 'train'
-_TEST = 'test'
-_TARGET_COLUMN = 'target_column'
-_TASK_TYPE = 'task_type'
+_DATA = "data"
```
Comment (imatiach-msft) [inline responsibleai/responsibleai/rai_insights/rai_insights.py:27]: these should stay as single quoted strings

File: responsibleai/responsibleai/rai_insights/rai_insights.py [hunk 31]
```diff
@@ -264,30 +323,33 @@ def _validate_model_analysis_input_parameters(
...
+            if classes is not None and task_type == ModelTask.CLASSIFICATION:
+                if (
+                    len(set(train[target_column].unique()) - set(classes)) != 0
+                    or len(set(classes) - set(train[target_column].unique()))
+                    != 0
+                ):
                     raise UserConfigValidationException(
-                        'The train labels and distinct values in '
-                        'target (train data) do not match')
+                        "The train labels and distinct values in "
```
Comment (imatiach-msft) [inline responsibleai/responsibleai/rai_insights/rai_insights.py:333]: I think this message would be clearer if written as:

```
The specified classes and distinct values in target (train data) do not match
```

File: libs/dataset-explorer/src/lib/DefaultTab.tsx (hunk 17 above)
Comment (kashmoneygt) [reply libs/dataset-explorer/src/lib/DefaultTab.tsx:106]: I renamed `DatasetExplorerTab` to `DefaultTab` and Git picked up as a completely new file instead of a file rename. So I did not change any code inside the Dataset Explorer :)

File: responsibleai/responsibleai/_internal/constants.py (hunk 22 above)
Comment (kashmoneygt) [reply responsibleai/responsibleai/_internal/constants.py:10]: Yeah, it's mainly because I'm using the `black` formatter which tries to convert single to double quotes where possible

I'll unconvert these for sure, but I'd highly recommend the `black` formatter for the Python code :)

https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html

File: CONTRIBUTING.md (hunk 2 above)
Comment (romanlutz) [reply CONTRIBUTING.md:51]: Shameless plug.... #LutzsLaw supports this statement 😉 https://twitter.com/romanlutz13/status/1451289545627086849?s=20&t=o8q06I1vrct9LJd_Ixx6LQ

File: libs/dataset-explorer/src/lib/DefaultTab.tsx [hunk 32]
```diff
@@ -0,0 +1,389 @@
...
+
+export interface IDefaultTabState {
+  xDialogOpen: boolean;
+  yDialogOpen: boolean;
+  colorDialogOpen: boolean;
+  selectedCohortIndex: number;
+  chartProps?: IGenericChartProps;
+}
+
+export class DefaultTab extends React.Component<
```
Comment (romanlutz) [inline libs/dataset-explorer/src/lib/DefaultTab.tsx:44]: Can we find a more descriptive name? I have no clue what to expect from `DefaultTab`.
